yoyodyne/data/tsv.py

```diff
@@ -0,0 +1,175 @@
+"""TSV parsing.
+
+The TsvParser yield string tuples from TSV files using 1-based indexing.
+
+The CellParser converts between raw strings ("cells") and lists of string
+symbols.
+"""
+
+import csv
+import dataclasses
+from typing import Iterator, List, Tuple
+
+from .. import defaults, util
+
+
+class Error(Exception):
+    """Module-specific exception."""
+
+    pass
+
+
+@dataclasses.dataclass
+class TsvParser:
+    """Streams cells from a TSV file.
+
+    Args:
+        source_col (int, optional): 1-indexed column in TSV containing
+            source strings.
+        features_col (int, optional): 1-indexed column in TSV containing
+            features strings.
+        target_col (int, optional): 1-indexed column in TSV containing
+            target strings.
+    """
+
+    source_col: int = defaults.SOURCE_COL
+    features_col: int = defaults.FEATURES_COL
+    target_col: int = defaults.TARGET_COL
+
+    # This is automatically called after initialization.
+    def __post_init__(self) -> None:
+        if self.features_col < 0:
+            raise Error(f"Invalid features column: {self.features_col}")
+        if self.features_col != 0:
+            util.log_info("Including features")
+        if self.source_col < 1:
+            raise Error(f"Invalid source column: {self.source_col}")
+        if self.target_col < 0:
+            raise Error(f"Invalid target column: {self.target_col}")
+        if self.target_col == 0:
+            util.log_info("Ignoring targets in input")
+
+    @staticmethod
+    def _tsv_reader(path: str) -> Iterator[str]:
+        with open(path, "r") as source:
```

Comment:
Not really important, but `source` is semantically overloaded in data modules (the file, or the source string).